Tracing: Various fixes and optimizations

nashif opened the pull request:
- Align all trace points for switching thread in/out on all architectures
- trace mutex/semaphore using dedicated calls
- ztest: set thread name
- trace k_sleep

nashif committed:
tracing: swap: cleanup trace points and their location

Move tracing switched_in and switched_out to the architecture code and
remove duplications. This changes swap tracing for x86, xtensa.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: arc: depend on CONFIG_TRACING_ISR for ISRs

Use CONFIG_TRACING_ISR to exclude tracing ISRs just like other
architectures.

Also, z_sys_trace_isr_exit was not defined (It was renamed some time ago
and this was forgotten...)

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: trace mutex/semaphore using dedicated calls

Instead of using generic trace calls, use dedicated functions for
tracing sempahores and mutexes.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: trace k_sleep

Trace when k_sleep is called.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: nios2: fix swap tracing

Fixed ordering of tracepoints for the nios2 architecture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: cortex_a53: fix order of swap tracing

We had switched_in and switched_out mixed up.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: riscv: fix tracing of context switch

We had switched_in and switched_out mixed up.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: posix_arch: trace swap

Moving trace points to the architecture code.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ztest: set thread name

If thread names are enabled, set thread name for the ztest thread for
tracing purposes.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: make uart backend sync capable

This works just fine, no need to make this backend async only.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: convert ctf timestamp to ns

We have been using cycles instead of time units.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tracing: ctf: capture thread names in ctf payload

Add thread name to the payload where applicable.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

andrewboie commented on arch/arc/core/switch.S:
```diff
@@ -132,6 +132,17 @@ _switch_return_from_coop:
 
 	pop_s blink
 #endif /* CONFIG_EXECUTION_BENCHMARKING */
+#ifdef CONFIG_TRACING
```
not sure I understand why this change was made.

It seems far better to me to do this in the core kernel, around invocations of `arch_switch()` or `arch_swap()` appropriately.

I really like to avoid doing stuff in arch code if we can do it in the core kernel, x1000 if doing in the arch layer requires assembly!

andrewboie replied:
Did you have to do this because of https://github.com/zephyrproject-rtos/zephyr/issues/20671 ?

nashif replied:
no, I switched trace point to C but this code was left by mistake. fixed now.

andrewboie commented on include/arch/arc/v2/irq.h:
```diff
@@ -92,9 +92,6 @@ extern void z_irq_spurious(void *unused);
 
 static inline void arch_isr_direct_header(void)
 {
-#ifdef CONFIG_TRACING
```
shouldn't this be changed to `CONFIG_TRACING_ISR` instead of completely removed? "direct" interrupts skip the common interrupt handling code AFAIK

nashif replied:
yes

andrewboie commented on kernel/include/kswap.h:
```diff
@@ -117,6 +117,7 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 		wait_for_switch(new_thread);
 		arch_switch(new_thread->switch_handle,
 			     &old_thread->switch_handle);
+		sys_trace_thread_switched_in();
```
commit message mentions x86_64, but this is in common code and applies to all arches using `CONFIG_SWITCH`

andrewboie commented on include/tracing/tracing.h:
```diff
@@ -132,6 +132,41 @@
+ * @brief Trace initialisation of a Semaphore
+ * @param sem Semaphore object
+ */
+#define sys_trace_semaphore_init(sem)
```
what do these empty definitions do? is this for when tracing is disabled? I don't think we would need doxygen here.

nashif replied:
this for when tracing is disabled, doxygen actually belongs here and not in the implementation.

tyrantlizard77 reviewed (commented):
Main concern is context switching trace points in arch asm code.
It may be the case that systems that use `arch_swap()` this is unavoidable due to how `arch_swap()` itself gets called from interrupt exit paths, but unless I've missed a detail somewhere, any `arch_switch()` system should be able to instrument this entirely in the core kernel.

ioannisg commented on arch/arm/core/aarch64/swap_helper.S:
```diff
@@ -35,7 +35,7 @@ GTEXT(z_arm64_context_switch)
 #ifdef CONFIG_TRACING
 	stp	xzr, x30, [sp, #-16]!
-	bl	sys_trace_thread_switched_in
+	bl	sys_trace_thread_switched_out
```
This looks like it was a bug.

ioannisg replied:
@carlocaione @sandeepbrcm ^^

nashif replied:
yes, it was a bug..

carlocaione replied:
indeed

ioannisg commented on include/arch/arc/v2/irq.h:
```diff
@@ -35,6 +35,8 @@ extern void z_arc_firq_stack_set(void);
 extern void arch_irq_enable(unsigned int irq);
 extern void arch_irq_disable(unsigned int irq);
 extern int arch_irq_is_enabled(unsigned int irq);
+extern void sys_trace_isr_enter(void);
```
should't these be guarded with the corresponding ifdef?

nashif replied:
they can be, sure.

jenmwms commented on arch/x86/core/ia32/swap.S:
```diff
@@ -87,6 +87,19 @@ SECTION_FUNC(TEXT, arch_swap)
+	call	sys_trace_thread_switched_out
+
+	popl	%edx
+	popl	%eax
```
Can you please help me understand why we add `push/pop` of `%eax` here? I know it is another register (for x86), and you mention we want to keep `isr` and `isr_param`, which would need 2 registers. How did you know to use `%eax`?

nashif replied:
> These same complaints apply to the `CONFIG_EXECUTION_BENCHMARKING` block immediately before it that is already in master branch. Which also brings up the question on why we have two instrumentation points right after one another.

yep, this is going away, there will be only one instrumentation point and we will use tracing for measuring timing for the benchmarks we have right now.

andrewboie commented on arch/x86/core/ia32/swap.S:
```diff
@@ -363,6 +369,17 @@ CROHandlingDone:
+
+	/* Save these as we are using to keep track of isr and isr_param */
+	pushl	%eax
+	pushl	%edx
```
we do need to preserve `arch_swap()`'s return value in `%eax`, but why are we pushing/popping `%edx`

Also editorial complaint that we have two different instrumentation points right after one another. What's the difference between `CONFIG_EXECUTION_BENCHMARKING` and `CONFIG_TRACING`? Can we unify?

nashif replied:
> Also editorial complaint that we have two different instrumentation points right after one another. What's the difference between `CONFIG_EXECUTION_BENCHMARKING` and `CONFIG_TRACING`? Can we unify? Maybe we can do this in a follow-up PR but it sticks out

not mentioned anywhere here but the idea is to drop CONFIG_EXECUTION_BENCHMARKING completely, been working with Daniel on this already and showing progress. This is just the first step toward this

andrewboie commented on arch/x86/core/ia32/swap.S:
```diff
@@ -363,6 +369,17 @@ CROHandlingDone:
+#endif
+#if defined(CONFIG_TRACING)
+
+	/* Save these as we are using to keep track of isr and isr_param */
```
this comment is wrong.

copypasta?

nashif replied:
yeah, copy/paste, let me fix that.

andrewboie commented on include/arch/arc/v2/irq.h:
```diff
@@ -104,8 +106,8 @@ static inline void arch_isr_direct_footer(int maybe_swap)
-#ifdef CONFIG_TRACING
-	z_sys_trace_isr_exit();
+#ifdef CONFIG_TRACING_ISR
+	sys_trace_isr_exit();
```
at least for `sys_trace_thread` and `sys_trace_sem` functions, we have them compile to nothing if tracing is disabled.

can we do the same for ISR, so we don't have to wrap these invocations in ifdefs?

nashif replied:
yeah, i dont like the too many ifdefs either, this will move the ifdef into the function however, which I think is fine..

andrewboie commented on include/tracing/tracing.h:
```diff
@@ -131,6 +135,41 @@
+ * @brief Trace taking a Semaphore
+ * @param sem Semaphore object
+ */
+#define sys_trace_semaphore_take(sem)
```
why are these preprocessor `#defines`?

these should be empty inline functions for type safety

never use a macro when an inline function will suffice IMO

nashif replied:
if I remember correctly we used to have them inline initially or at least in the  original PR from 2-3 years ago, some users however wanted them to be defines to make it easy to integrate with existing tracing formats. There was some other complication I do not recall, need to check and will revisit, in any case, this would need to happen in a different PR.

jenmwms commented on kernel/include/kswap.h:
```diff
@@ -113,11 +112,11 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 		wait_for_switch(new_thread);
 		arch_switch(new_thread->switch_handle,
 			     &old_thread->switch_handle);
-		sys_trace_thread_switched_in();
```
I'm curious, was this part of the problem for counting switched in twice? It looks like it, but wanted to ask.

nashif replied:
this is added again in a later patch, they will be sqaushed, but yes, we were tracing both in C and asm.

nashif committed:
tracing: x86: trace isr_exit

We were missing exit from ISR..

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

jenmwms commented on include/tracing/tracing.h:
```diff
@@ -50,16 +50,19 @@
 /**
  * @brief Called when a thread is being aborted
+ * @param thread Thread structure
  *
```
nit: extra line in comment

nashif committed:
tracing: scripts: add scripts for parsing ctf data

Generate trace using samples/subsys/tracing for example:

    west build -b qemu_x86 samples/subsys/tracing  -t run \
      -- -DCONF_FILE=prj_uart_ctf.conf

    mkdir ctf
    cp build/channel0_0 ctf/
    cp subsys/tracing/ctf/tsdl/metadata ctf/
    ./scripts/tracing/parse_ctf.py -t ctf

1969-12-31 19:00:00.001779 (+0.000000 s): thread_create: thread_a
1969-12-31 19:00:00.001915 (+0.000136 s): thread_info (Stack size: 1024)
1969-12-31 19:00:00.002021 (+0.000107 s): thread_ready: thread_a
1969-12-31 19:00:00.002287 (+0.000265 s): thread_switched_out: main
1969-12-31 19:00:00.002320 (+0.000033 s): thread_switched_in: main
1969-12-31 19:00:00.002585 (+0.000265 s): thread_abort: main
1969-12-31 19:00:00.002730 (+0.000145 s): thread_switched_out: main
1969-12-31 19:00:00.002762 (+0.000032 s): thread_switched_in: thread_a
1969-12-31 19:00:00.002829 (+0.000067 s): thread_create: 1130656
1969-12-31 19:00:00.002860 (+0.000031 s): thread_info (Stack size: 1024)
1969-12-31 19:00:00.002911 (+0.000052 s): thread_ready: 1130656
1969-12-31 19:00:00.003033 (+0.000121 s): thread_name_set
1969-12-31 19:00:00.003132 (+0.000100 s): semaphore_take (1140992)
1969-12-31 19:00:00.003201 (+0.000069 s): end_call 38 (SEMA_TAKE)
1969-12-31 19:00:00.003330 (+0.000128 s): start_call 39 (SLEEP)
1969-12-31 19:00:00.003721 (+0.000391 s): thread_suspend: thread_a
1969-12-31 19:00:00.003802 (+0.000081 s): thread_switched_out: thread_a
1969-12-31 19:00:00.003834 (+0.000032 s): thread_switched_in: thread_b
1969-12-31 19:00:00.003872 (+0.000038 s): semaphore_take (1141016)
1969-12-31 19:00:00.003976 (+0.000103 s): thread_pending: thread_b
1969-12-31 19:00:00.004095 (+0.000119 s): thread_switched_out: thread_b
1969-12-31 19:00:00.004127 (+0.000032 s): thread_switched_in: idle 00
1969-12-31 19:00:00.004232 (+0.000105 s): idle
1969-12-31 19:00:00.510848 (+0.506616 s): isr_enter
1969-12-31 19:00:00.511664 (+0.000816 s): thread_resume: thread_a
1969-12-31 19:00:00.511947 (+0.000283 s): thread_ready: thread_a
1969-12-31 19:00:00.512385 (+0.000438 s): isr_exit
1969-12-31 19:00:00.512570 (+0.000186 s): thread_switched_out: idle 00
1969-12-31 19:00:00.512689 (+0.000119 s): thread_switched_in: thread_a
1969-12-31 19:00:00.512869 (+0.000180 s): end_call 39 (SLEEP)

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

jenmwms commented on kernel/sem.c:
```diff
@@ -69,9 +69,9 @@ int z_impl_k_sem_init(struct k_sem *sem, unsigned int initial_count,
-	sys_trace_void(SYS_TRACE_ID_SEMA_INIT);
 	sem->count = initial_count;
 	sem->limit = limit;
+	sys_trace_semaphore_init(sem);
```
I understand the change to `sys_trace_semaphore_init(sem)`. Why did we move it below the `sem-`>` assignments? In this position, does it better represent the timing of the trace at initializing the semaphore (as in the next line z_waitq_init)?

nashif replied:
> Why did we move it below the `sem-`>` assignments?

to get the values the semaphore was initiated with, this could be part of the tracing later.

jenmwms commented on subsys/tracing/include/tracing_test.h:
```diff
@@ -31,7 +31,12 @@ void sys_trace_isr_exit_to_scheduler(void);
 void sys_trace_void(unsigned int id);
 void sys_trace_end_call(unsigned int id);
-
+void sys_trace_semaphore_init(struct k_sem *sem);
```
jw - I expected to see `sys_trace_start_call()` here (and ctf_top.c/h), too similar to the sys_trace_end_call, etc.. Looks like it hasn't been there yet but things work. Is it missing? Please correct me if I'm wrong.

nashif replied:
yeah, looks like sys_trace_start_call is missing in the test variant, this is just a test, no harm :)

jenmwms commented on arch/x86/core/ia32/intstub.S:
```diff
@@ -203,6 +200,12 @@ alreadyOnIntStack:
 
+#if defined(CONFIG_TRACING_ISR)
+	pushl	%eax
+	call	sys_trace_isr_exit
```
Does this need the `pushl/popl %edx` like above in line 183?

andrewboie replied:
no I don't believe so, it's unused after the call statement, we even zero it on line 211 if X2APIC is active

but we also don't need to push/pop EAX here either, note how it is zeroed on line 209

jenmwms commented on scripts/tracing/parse_ctf.py:
```diff
@@ -0,0 +1,151 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+"""
+Script to parse CTF data and print to the screen in a custom and colorful
```
Do I understand the role of this script? This script looks specific to babeltrace. Does it use babeltrace2 to interpret the CTF channel and metadata, and this script works to customize the display using babeltrace APIs?

nashif replied:
this is just a helper script to read the generated data on the command line.

andrewboie commented on arch/x86/core/ia32/intstub.S:
```diff
@@ -192,6 +179,16 @@ alreadyOnIntStack:
+
+#if defined(CONFIG_TRACING_ISR)
+	/* Save these as we are using to keep track of isr and isr_param */
+	pushl	%eax
```
I don't think EAX needs to be pushed here, it used to contain the interrupt handler argument but that got pushed onto the stack earlier

The CONFIG_EXECUTION_BENCHMARKING block is also wrong but that's eventually being removed.

The stench of copy-paste is strong here...have to be very careful about doing this in assembly as there is no compiler to fix redundancies

abrodkin reviewed (approved):
LGTM
